lib/vmpooler/metrics/promstats/collector_middleware.rb

```diff
@@ -0,0 +1,119 @@
+      # Label Builder functions will receive a Rack env and a status code, and must
+      # return a hash with the labels for that request. They must also accept an empty
+      # env, and return a hash with the correct keys. This is necessary to initialize
+      # the metrics with the correct set of labels.
+      class CollectorMiddleware
+        attr_reader :app, :registry
+
+        def initialize(app, options = {})
+          @app = app
+          @registry = options[:registry] || Prometheus::Client.registry
+          @metrics_prefix = options[:metrics_prefix] || 'http_server'
+
+          init_request_metrics
+          init_exception_metrics
+        end
+
+        def call(env) # :nodoc:
+          trace(env) { @app.call(env) }
+        end
+
+        protected
+
+        def init_request_metrics
+          @requests = @registry.counter(
+            :"#{@metrics_prefix}_requests_total",
+            docstring:
+              'The total number of HTTP requests handled by the Rack application.',
+            labels: %i[code method path]
+          )
+          @durations = @registry.histogram(
+            :"#{@metrics_prefix}_request_duration_seconds",
+            docstring: 'The HTTP response duration of the Rack application.',
+            labels: %i[method path]
+          )
+        end
+
+        def init_exception_metrics
+          @exceptions = @registry.counter(
+            :"#{@metrics_prefix}_exceptions_total",
+            docstring: 'The total number of exceptions raised by the Rack application.',
+            labels: [:exception]
+          )
+        end
+
+        def trace(env)
+          response = nil
+          duration = Benchmark.realtime { response = yield }
+          record(env, response.first.to_s, duration)
+          return response
+        rescue => exception
+          @exceptions.increment(labels: { exception: exception.class.name })
+          raise
+        end
+
+        def record(env, code, duration)
+          counter_labels = {
+            code:   code,
+            method: env['REQUEST_METHOD'].downcase,
+            path:   strip_ids_from_path(env['PATH_INFO']),
+          }
+
+          duration_labels = {
+            method: env['REQUEST_METHOD'].downcase,
+            path:   strip_ids_from_path(env['PATH_INFO']),
+          }
+
+          @requests.increment(labels: counter_labels)
+          @durations.observe(duration, labels: duration_labels)
+        rescue
+          # TODO: log unexpected exception during request recording
+          nil
+        end
+
+        def strip_ids_from_path(path)
+            # Custom for /vm path - so we just collect aggrate stats for all usage along this one
+            # path. Custom counters are then added more specific endpoints in v1.rb
+            path
+            .gsub(%r{/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}(/|$)}, '/:uuid\\1')
+            .gsub(%r{/\d+(/|$)}, '/:id\\1')
+            .gsub(%r{/vm/.+$}, '/vm/:host')
```

Comment:
How does this output for templates and such (aka, non-hosts)?

Reply:
Fixed as per discussion